### ISSUE TITLE
m.component Semantics Proposal

### ISSUE THREAD

**Issue by mindeavor** [lhorie/mithril.js#530]
Based on our discussion in #497, I've created a [wiki page](https://github.com/lhorie/mithril.js/wiki/Components-Syntax-Proposal) describing proposed changes to the semantics of what `m.component` would be. Please discuss and we can update the wiki if we find flaws / better solutions.

**Comment by lhorie** [comment]
I committed the documentation files for the renamed API points ([`m.mount`](https://github.com/lhorie/mithril.js/blob/components/docs/mithril.mount.md) and [`m.component`](https://github.com/lhorie/mithril.js/blob/components/docs/mithril.component.md)). The prose is still work in progress, but the signature and exaples reflect the implementation correctly.

**Comment by mindeavor** [comment]
For those following, here are the links @lhorie is talking about:

- [m.component docs](https://github.com/lhorie/mithril.js/blob/components/docs/mithril.component.md)
- [m.mount docs](https://github.com/lhorie/mithril.js/blob/components/docs/mithril.mount.md)

These docs reflect what is currently in the components branch. The wiki page I created is an aggregation of the discussion from #497

**Comment by tobscure** [comment]
Just been playing around and found that the `var MyComponent = m.component({ controller: ..., view: ... })` approach introduces a problem: you can't really alter the original component object after you've wrapped it. (I'm building an extensible app for which this is important.)

```javascript
// old way, what I need to be able to do
var OldComponent = { controller: ..., view: ... }
OldComponent.controller.prototype.foo = 'bar'

// this would be fine, but this syntax currently isn't working
m('div', m.component(OldComponent, {arg: 'whatever'}))

// not as nice imo
m('div', m.component(OldComponent)({arg: 'whatever'}))

// new way
var NewComponent = m.component({ controller: ..., view: ... })

// only alters this instance of NewComponent(), not subsequent ones
NewComponent().controller.prototype.foo = 'bar'

// I think the original controller can be accessed like this,
// although obviously this is undefined/undocumented behaviour
NewComponent().controller.$original
```

So I feel like pre-wrapping a component with m.component adds an unnecessary layer of complexity and it's slightly harder to reason about what exactly the `NewComponent` variable actually represents. I think part of the confusion is that if m.component returns a function (not a constructor), then its variable assignment shouldn't be named like a constructor (i.e. myComponent instead of MyComponent)

Personally I think the `m('div', m.component(OldComponent, {arg: 'whatever'}))` syntax is a lot more intuitive, but as I mentioned, this is currently broken :)

**Comment by tobscure** [comment]
What about a syntax like this?

```javascript
var MyComponent = function(args) {
  this.foo = args.foo
  this.bar = args.bar

  this.view = function(ctrl) {
    return [
      m('p', this.foo),
      m('p', this.bar)
    ]
  }
}
MyComponent.controller = function(component) {
  console.log(component.href, component.label)
}

var component = new MyComponent({
  foo: 'Hello',
  bar: 'World!'
})

m.mount(document.body, component)
```

**Comment by webcss** [comment]
@tobscure I think you're doing it wrong
```javascript
// old way, what I need to be able to do
var OldComponent = { controller: ..., view: ... }
OldComponent.controller.prototype.foo = 'bar'

// you don't use OldComponent like this
m('div', m.component(OldComponent, {arg: 'whatever'}))

// but rather like this
m('div', OldComponent( {args: 'whatever'} )
```

m.component is only for component declaration.

**Comment by farzher** [comment]
How does mithril know to call the controller for each component only once? (I'm assuming that's how it works)

**Comment by a-s-o** [comment]
Hello everyone. I am big fan of Mithril and I have been following the component discussions here and in the components docs eagerly waiting for the release. I use the components/module pattern extensively in my site.

At the risk of obscuring the issue at hand, I would like to share a little pattern I have been using to manage stateless and stateful components. I have two helper methods (seem similar to the new mount and component methods being proposed:
```
   m.createComponent = function (comp, params) {
      var Controller = comp.controller || function() {};
      var View = comp.view || function() {};

      // All component controllers accept a single argument (hint: use objects)
      var instance, boundArgs;
      if (arguments.length > 1) {
         params = params || {};
         boundArgs = [params];
         instance = new Controller(params);
      } else {
         boundArgs = [];
         instance = new Controller();
      }

      // Return a new object with some additional functions,
      // to make rendering and loading easier. The controller instance
      // is available at the instance property
      return {
         controller : comp.controller,
         view       : comp.view,
         instance   : instance,
         render     : function (...moreArgs) {
            return View.apply(null, [instance].concat(boundArgs, moreArgs));
         },
         onunload  : function () {
            if (instance.onunload) instance.onunload();
         }
      };
   };

   // Simple components can simple use this method to render directly
   m.renderComponent = function (comp, params, ...moreArgs) {
      if (arguments.length < 2) return m.createComponent(comp).render();
      return m.createComponent(comp, params).render.apply(null, moreArgs);
   };
```
Both methods accept simple components `{ controller: fn, view: fn  }` as first argument and params as a second argument. I use renderComponent() directly in the my consumer views when I do not need a reference to my controller. Ex:
```
var Button = require('component/button');
someView: function (ctrl, params = {}) {
   return m('div.LoginForm', [
      m(Button, { buttonOptions: {}, onclick: params.binds }, 'Click to log in')
   ])
}
```
I use createComponent() for more complicated components. Ex: 
```
var ListEditor = require('component/list-editor');
var consumer = {
   controller: function (params = {}) {
      var ctrl = this;
      ctrl.list = m.createComponent(ListEditor, { list: params.list.slice() });
      ctrl.save = function () {
         m.request({ method: 'POST', url: '/save-list', data: ctrl.list.instance.toJSON() });
      }
      ctrl.onunload = function () {
         ctrl.list.onunload();
      }
   },
   view: function (ctrl, params = {}, additionalArgs = {}) {
      return [
           ctrl.list.render(), // ListEditor view will be rendered with args provided in controller
           m.createComponent(Button, { onclick: ctrl.save })
      ]
   }
}
```
And that's it for majority of the cases I have encountered. I have not seen any performance or memory leak type issues in about a month or so of use. 

I believe I saw something similar to this in these discussions and then just tweaked to my preference. These methods don't solve the issue at hand of auto-instantiating and auto-unloading components but make the lifecycle a little easier for pure components (which also nudges me to write more of those).

**Comment by a-s-o** [comment]
Whoa, hold the phone. I was just thinking about my post right above here and reading m.component docs and came up with this slightly less wordy alternative.
```
function componentize (comp) {
   var View = comp.view || function() {};
   var Controller = comp.controller || function() {};
   
   return function Component (params, ...args) {
      var instance, boundArgs;
      if (arguments.length > 1) {
         params = params || {};
         boundArgs = [params];
         instance = new Controller(params);
      } else {
         boundArgs = [];
         instance = new Controller();
      }

      // When called with new keyword, return an object, otherwise return a rendered view
      if (this instanceof Component ) {
         return {
            controller : comp.controller,
            view       : comp.view,
            instance   : instance,
            render     : function (...moreArgs) {
               return View.apply(null, [instance, ...boundArgs, ...moreArgs]);
            },
            onunload  : function () {
               if (instance.onunload) instance.onunload();
            }
         };
      } else {
         return View.apply(null, [instance, params, ...args]);
      }
   };
}
```
A little experimental, as I don't know enough about javascript to know whether there are some edge case where using `new` or `instanceof` checks in this way may be unsafe. Please let me know.  
```
// Usage: Alternative 1 - simple components
var Button = componentize({ controller: function () {}, view: function () {} });
someView: function (ctrl, params = {}) {
   var buttonParams = { buttonOptions: {}, onclick: params.binds };
   return m('div.LoginForm', [
      Button(buttonParams, 'Click to log in')
   ])
}

// Usage: Alternative 2 - stateful components
var ListEditor = componentize({ controller: function () {}, view: function () {} });
var consumer = {
   controller: function (params = {}) {
      var ctrl = this;
      
      // Use the component with new when access to controller or onunload is required
      ctrl.list = new ListEditor({ list: params.list.slice() });

      ctrl.save = function () {
         m.request({ 
            method: 'POST', url: '/save-list',
            // Get the data from the list component 
            data: ctrl.list.instance.toJSON() 
         });
      }
      ctrl.onunload = function () {
         ctrl.list.onunload();
      }
   },
   view: function (ctrl, params = {}, additionalArgs = {}) {
      return [
           // ListEditor view will be rendered with args provided in controller
           ctrl.list.render({ color: additionalArgs.color }), 
           m.createComponent(Button, { onclick: ctrl.save })
      ]
   }
}

**Comment by barneycarroll** [comment]
@a-s-o could you demonstrate your implementation as a fiddle / codepen / jsbin? Always interested in thinking about alternative methods, but it's a lot easier to analyse proposals in their own right when they're presented with use case code somewhere where you can consider them in their own right.

**Comment by a-s-o** [comment]
Done, please see 
createComponent/renderComponent style: http://jsbin.com/bepenayuza/2/
m.component style: http://jsbin.com/hewosunite/2/

**Comment by mindeavor** [comment]
First impressions: nice idea, very clever. I have to say I do like the idea
of having the option to use components in these two different ways. Before
the components branch came along, I was making good use of a style similar
to createComponent / renderComponent.

**Comment by lhorie** [comment]
@tobscure I might just be missing the point, but I believe you can still use what you call the old way, e.g.

```javacript
var OldComponent = { controller: ..., view: ... }
OldComponent.controller.prototype.foo = 'bar'

//then use as usual
var MyComponent = m.component(OldComponent)

m("div", MyComponent({baz: "quux"}))
```

@a-s-o one of the main use cases that the components branch is aiming to address is list management (or more generically, the ability to scale identity management complexity-wise), e.g. 

```javascript
m("ul", [
  someList.map(function(item) {
    return MyComponent({id: item.id, onremove: item.remove}) //no need to add boilerplate code to manage list of component instances
  })
])
```

If that aspect isn't important to your particular case, and you're only interested in components for their organizational aspects, recall that components are still backwards compatible, so you can still use the old componentization style:

```javascript
//in controller
this.foo = new Foo.controller

//in view
Foo.view(ctrl.foo)
```

And of course you could always deviate and do your own thing:

```javascript
//define a custom thing
var MyThing = function() {
  this.render = function() {
      return m("div", hello)
  }
}

//in another controller
this.foo = new MyThing

//in view
m("div", ctrl.foo.render())
```

**Comment by a-s-o** [comment]
Oh I see, so m.component handles that as well. That is fantastic. My intention is to just share patterns that I am using to see if they me assist in the semantics discussion

**Comment by mindeavor** [comment]
@lhorie I think the components API is ok without the arguments preloading/overriding stuff in the wiki. However, I think there is one main thing left to discuss: **Will the use of `m.component` be required to include a component inside another component's view?**

If no, then that means plain objects as contents (e.g. stuff in the array in `m('div', [ ... ])`) will continue to work as long as they have a `controller` and/or `view` property, correct?

If yes, then would that pave the way to solve issues with the first component caveat [listed in the docs](https://github.com/lhorie/mithril.js/blob/components/docs/mithril.component.md#nested-component-limitations-and-caveats), considering you could "tag" the object inside the `m.component` function?

**Comment by lhorie** [comment]
@mindeavor anything that looks like `{controller, view}` is considered a component (actually, simply having a view is enough, since controller is optional), so, no, a component doesn't necessarily need to be created via `m.component`. Technically, you only need to use `m.component` to be able to parameterize a component, but in the docs, I'm making all component declarations call `m.component` for consistency (to avoid confusion).

WRT to the limitation of a root element needing to be an element: There is one caveat to that caveat that is currently wrong in the documentation: a component may be the root element of another component. However, adding a `tag` property to a component object won't create an implicit tag (and I don't think allowing "hiding" of a tag like this would be a good idea)

**Comment by mindeavor** [comment]
@lhorie Makes sense, thanks. As it currently stands, would "late" style components have to look like this when parameterizing?

```javascript
var MyWidget = {
  view: function (ctrl, args) {
    return m('div', args.test)
  }
})

m.mount(document.body, {
  view: function () {
    return m.component(MyWidget)({ test: 10 })
  }
})
```

**Comment by lhorie** [comment]
Well, ideally, they'd look more like this

```javascript
var widget = m.component(MyWidget)

m.mount(document.body, {
  view: function() {
    return widget({test: 10})
  }
})
```

Inlining the `m.component` call technically works, but it's kinda ugly...

**Comment by mindeavor** [comment]
Yeah, I see what you mean. I'm just trying to think of all the styles people would use `m.component`:

### Upfront Style

```javascript
var MyWidget = m.component({
  controller: function() {},
  view: function (ctrl, args) { return m('div', args.test) }
})
m.mount(document.body, {
  view: function () {
    return MyWidget({ test: 10 })
  }
})
```

### Late Style

```javascript
var MyWidget = {
  controller: function() {},
  view: function (ctrl, args) { return m('div', args.test) }
}
m.mount(document.body, {
  view: function () {
    return m.component(MyWidget)({ test: 10 })
  }
})
```

### Assignment Style

```javascript
var MyWidget = m.component()

MyWidget.controller = function() {}
MyWidget.view = function (ctrl, args) { return m('div', args.test) }

m.mount(document.body, {
  view: function () {
    return MyWidget({ test: 10 })
  }
})
```

Personally I use the assignment style, so if this is officially supported (i.e. compatible with future versions), I'll be happy.

**Comment by lhorie** [comment]
Assignment style is actually not supported by the code, currently
```

**Comment by tobscure** [comment]
Will this syntax (the way it used to work with m.module) be supported?

```javascript
var MyComponent = {
  view: function(ctrl, args) { return m('div', args.test) }
}

m.mount(document.body, {
  view: function () {
    return m.component(MyComponent, { test: 10 })
  }
})
```

**Comment by lhorie** [comment]
@tobscure no

**Comment by farzher** [comment]
@lhorie Why not? Consistency?

**Comment by lhorie** [comment]
@farzher yes, type consistency. Basically, you shouldn't have to second guess whether a component you just `require`d can be called as a function or not. If a module (in the ES6/AMD/CommonJS sense) exposes a factory, it's callable, if it exposes a component object, it's an object.

The documentation will recommend exposing the factory, i.e. `module.exports = m.component(...)`

**Comment by rrag** [comment]
Can a component be `{controller, view, attrs, children}` and not just `{controller, view}`. A `m('div', {}, {})` is `{tag, attrs, children}` I understand component does not have a `tag` but it seems logical to include `attrs, children`

**Comment by lhorie** [comment]
@rrag you can pass those as arguments at invocation time, e.g. 

```
var Foo = m.component(...)

m("div", [
  Foo({id: 1}, "hello")
])
```

Note that the arguments aren't auto-bound to the root element (with the exception of `key`) to allows a bit more flexibility

**Comment by rrag** [comment]
@lhorie I realize that. I already asked this question in the google groups, here is a [fiddle](http://jsfiddle.net/rrag/58ezLg5m/26/) to illustrate my usecase. 

I want all options 1,2,... 6 to have the same `name`, it is possible for me to do that if I use a tag, but not when using a component. This gives the ability for a parent to change the `attrs` and `children` of its children and grand children. I see this as equivalents of [React.Children](http://facebook.github.io/react/docs/top-level-api.html#react.children)

**Comment by mindeavor** [comment]
@lhorie Besides being nice to look at, supporting "assignment style" is important for breaking up components into multiple files. I have an older project running 0.2.x on Rails that has the controller in a `coffee` file, and the view in an `msx` file. I'm in the process of refactoring it to use the latest components branch. Here's what I have to do at the moment:

```coffeescript
# user-profile-ctrl.coffee
window.UserProfile ||= {}
UserProfile.controller = (options) -> #...
```

```jsx
// user-profile-view.msx
window.UserProfile || (window.UserProfile = {})
UserProfile.view = function (ctrl) {
  return <div class="user-profile">...</div>
}
```

```coffeescript
# app.coffee
App = {}
App.view = (ctrl) ->
  m('.app', [
    m.component(UserProfile)({ user: ctrl.user() })
  ])
```

Which isn't bad per se, but it'd be nice to either be able to use assignment style, or have syntax support for `m.component(UserProfile, { user: ctrl.user() })`. It's certainly not a deal breaker, I just want to show a real use case for my earlier suggestions.

**Comment by mindeavor** [comment]
I'm finding @a-s-o's idea of having two different ways to use `m.component` interesting. What do you guys think about this alternative?

```javascript
MyWidget = { controller: ..., view: ... }
App = {
  controller: function () {
    // Key point: when you use the keyword `new`...
    this.widget = new m.component(MyWidget, { arg: 10 })
    // ...you get access to the component instance!
    this.widget.instance.whatever
  },
  view: function (ctrl) {
    return m('.app', [
      // Plain render
      ctrl.widget.render(),
      // Advanced render
      // Note how you can render the same component multiple times!
      ctrl.widget.render({ additional: 'params' }, [
        m('p', "Maybe some content")
      ]),
      // On the other hand, without the keyword `new`,
      // Mithril won't have to care about making the instance available,
      // potentially saving on precious cpu cycles.
      m.component(MyWidget, { arg: 11 }, [
        m('p', "Maybe some content")
      ]),
      m.component(MyOtherWidget, { arg: 11 })
    ])
  }
}
```

**Comment by mindeavor** [comment]
So after some discussion in the chat room, I think the `new` part of my previous comment is not a common enough use case to go into core (here's [an implementation](https://gist.github.com/mindeavor/0291304f829f6e40d398#file-m-ext-js-L13) for those interested).

Re other points: The callable syntax is great, but I don't think it should be in core (for those that want it, they can [implement it themselves](https://gist.github.com/mindeavor/25b8aa6810b244665a2e). Mithril will be easier to teach and to learn if there is one official way to use subcomponents, and the style of `m.component(MyWidget, { arg: 11 })` is simple and easy to understand.

I have one suggestion for the semantics of `m.component` to make it more extensible and export-friendly, but it doesn't need to be in the initial release of components, so I'll create a separate issue later.

In short, the simplicity of `m.component(MyWidget, { arg: 11 })` is elegant enough to solidify the teaching, learning, and documentation of Mithril.

**Comment by lhorie** [comment]
So, just for the record, as @mindeavor mentioned, there was some discussion about the API in the chat room a few days ago about the components API and the gist is that I've been dogfooding the new API (while writing the docs and trying to use it in general), and I was running into a bunch of annoyances that I didn't run into before, so I've decided to change `m.component` back to how `m.module(component, args...)` worked before.

The docs in the branch should reflect that change.

I've been thinking about @rrag's request, and I think I won't implement it because AOP can be accomplished by using a plain virtual dom tree (as opposed to using components), or via config and the W3C API (aka by playing the free out-of-jail card). I feel that allowing components to magically modify other components violates their encapsulation principle. If one wants to deliberately do that, it ought to be either explicit or done at a low level (config)

The documentation currently discourages using config thatway, but I'be been finding instances where using config to modify other elements is needed (for example, to make drag-drop smooth)

**Comment by tropperstyle** [comment]
Hello,

I have been closely following the development of the components branch and I just wanted to voice my support of the switch back to `m.module({..}, args..)` - I felt that the `c.component({..})` declarations were a huge step backwards in terms of a lean, intuitive API. So again, just happy to see the project not move in that direction.

If anything, I prefer `m(module, args..)` to `m.module(module, args..)` as it is more consistent with `m(...)` being the only way to compose virtual elements.

**Comment by mindeavor** [comment]
@tropperstyle Others have suggested using `m(module, args)` in the past. The counter argument is the current `m` function returns a VDOM element, while `m(module, args)` would not. It would create the following inconsistency:

```javascript
m.route(document.body, '/a', {
  '/a': m(MyWidget, { x: 11}), // <-- Possible
  '/b': m('h3') // <-- Not possible
})
```

**Comment by lhorie** [comment]
Given that the components API is very unlikely to change anymore after the `m.component()()` misstep, I'm gonna go ahead and close this
